File: moveit_nodes_pkg/src/mover.cpp
#include <memory>
#include <vector>
#include <string>
#include <iostream>
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_interface/planning_interface.h>
#include <moveit_msgs/msg/robot_trajectory.hpp>
#include <moveit_msgs/srv/get_motion_plan.hpp>
#include <xarm_msgs/srv/mover_service.hpp>
#include <xarm_msgs/msg/xarm7_unity_target_pose_to_ros.hpp>
#include <xarm_msgs/msg/xarm7_ros_joints_to_unity.hpp>
#include <tf2_eigen/tf2_eigen.hpp>
#include <tf2_ros/transform_listener.h>
#include <Eigen/Geometry>



using namespace std::chrono_literals;
using MoverService = xarm_msgs::srv::MoverService;
using MoverServiceRequest = xarm_msgs::srv::MoverService::Request;
using MoverServiceResponse = xarm_msgs::srv::MoverService::Response;
using Xarm7UnityTargetPoseToROS = xarm_msgs::msg::Xarm7UnityTargetPoseToROS;
using Xarm7ROSJointsToUnity = xarm_msgs::msg::Xarm7ROSJointsToUnity;

static const rclcpp::Logger LOGGER = rclcpp::get_logger("MoverNode");

class MoverNode
{
public:
  MoverNode(rclcpp::Node::SharedPtr &node) : node_(node)
  {
    node_->declare_parameter<std::string>("group_name", "xarm7");

    std::string group_name;
    node_->get_parameter("group_name", group_name);

    init(group_name);

    pub_ = node_->create_publisher<Xarm7ROSJointsToUnity>("xarm_ros_joints_to_unity", 10);
    sub_ = node_->create_subscription<Xarm7UnityTargetPoseToROS>("xarm_unity_target_pose_to_ros", 10,
                                                                std::bind(&MoverNode::xarm7UnityCallback, this, std::placeholders::_1));
    srv_ = node_->create_service<MoverService>("xarm_moveit", std::bind(&MoverNode::planPickAndPlace, this,
                                                                       std::placeholders::_1, std::placeholders::_2));

    RCLCPP_INFO(LOGGER, "Ready to plan");
  }

  void init(const std::string &group_name)
  {
    is_trajectory_ = false;
    move_group_ = std::make_shared<moveit::planning_interface::MoveGroupInterface>(node_, group_name);
    RCLCPP_INFO(LOGGER, "Planning frame: %s", move_group_->getPlanningFrame().c_str());
    RCLCPP_INFO(LOGGER, "End effector link: %s", move_group_->getEndEffectorLink().c_str());
    RCLCPP_INFO(LOGGER, "Available Planning Groups:");
    std::copy(move_group_->getJointModelGroupNames().begin(), move_group_->getJointModelGroupNames().end(), std::ostream_iterator<std::string>(std::cout, ", "));
  }

private:
  void xarm7UnityCallback(const Xarm7UnityTargetPoseToROS::SharedPtr msg)
  {
    if (!is_trajectory_)
    {
      is_trajectory_ = true;
      try
      {
        motion_plan_ = planTrajectory(msg->target_pose, msg->robot_arm_rotation);
        sendPlanToUnity(motion_plan_.trajectory_);
      }
      catch (const std::exception &ex)
      {
        RCLCPP_ERROR(LOGGER, "Exception in planning: %s", ex.what());
        is_trajectory_ = false;
      }
    }
  }

  moveit::planning_interface::MoveGroupInterface::Plan planTrajectory(const geometry_msgs::msg::Pose &destination_pose,
                                                                      const geometry_msgs::msg::Vector3 &robot_arm_rotation)
  {
    // geometry_msgs::msg::Pose new_destination_pose = rotatePose(destination_pose, robot_arm_rotation.x, robot_arm_rotation.y, robot_arm_rotation.z);

    // move_group_->setPoseTarget(new_destination_pose);


    RCLCPP_INFO(LOGGER, "Destination Pose Position = [x: %f, y: %f z: %f]", 
                        destination_pose.position.x,
                        destination_pose.position.y,
                        destination_pose.position.z);
    move_group_->setPoseTarget(destination_pose);

    moveit::planning_interface::MoveGroupInterface::Plan plan;
    bool success = (move_group_->plan(plan) == moveit::core::MoveItErrorCode::SUCCESS);
  
    if (!success)
    {
      throw std::runtime_error("Trajectory could not be planned. Please ensure target and destination are reachable by the robot.");
    }else{
      // move_group_->execute(plan);
  }

    return plan;
  }

  

geometry_msgs::msg::Pose rotatePose(const geometry_msgs::msg::Pose &pose, double roll, double pitch, double yaw) {
  // Convert degrees to radians
  roll = roll * M_PI / 180.0;
  pitch = pitch * M_PI / 180.0;
  yaw = yaw * M_PI / 180.0;

  // Create the rotation matrix from roll, pitch, and yaw
  Eigen::Matrix3d rotation_matrix;
  rotation_matrix = Eigen::AngleAxisd(roll, Eigen::Vector3d::UnitX())
                  * Eigen::AngleAxisd(pitch, Eigen::Vector3d::UnitY())
                  * Eigen::AngleAxisd(yaw, Eigen::Vector3d::UnitZ());

  // Convert the current orientation (quaternion) to a rotation matrix
  Eigen::Quaterniond current_orientation(pose.orientation.w, pose.orientation.x, pose.orientation.y, pose.orientation.z);
  Eigen::Matrix3d current_rotation_matrix = current_orientation.toRotationMatrix();

  // Combine the rotation matrices
  Eigen::Matrix3d new_rotation_matrix = rotation_matrix * current_rotation_matrix;

  // Extract the new orientation as a quaternion
  Eigen::Quaterniond new_orientation(new_rotation_matrix);

  // Create the new pose with the rotated orientation
  geometry_msgs::msg::Pose new_pose;
  new_pose.position.x = pose.position.x;
  new_pose.position.y = pose.position.y;
  new_pose.position.z = pose.position.z;
  new_pose.orientation.x = new_orientation.x();
  new_pose.orientation.y = new_orientation.y();
  new_pose.orientation.z = new_orientation.z();
  new_pose.orientation.w = new_orientation.w();

  return new_pose;
}


  void planPickAndPlace(const MoverServiceRequest::SharedPtr req, MoverServiceResponse::SharedPtr res)
  {
    try
    {
      auto pre_grasp_pose = planTrajectory(req->pick_pose, req->robot_arm_rotation);
      if (pre_grasp_pose.trajectory_.joint_trajectory.points.empty())
        return;

      // sendPlanToUnity(pre_grasp_pose.trajectory_);
      auto previous_ending_joint_angles = pre_grasp_pose.trajectory_.joint_trajectory.points.back().positions;

      auto pick_pose = req->pick_pose;
      pick_pose.position.z -= 0.05; // Adjust node_ value as necessary

      auto grasp_pose = planTrajectory(pick_pose, req->robot_arm_rotation);
      if (grasp_pose.trajectory_.joint_trajectory.points.empty())
        return;

      // sendPlanToUnity(grasp_pose.trajectory_);
      previous_ending_joint_angles = grasp_pose.trajectory_.joint_trajectory.points.back().positions;

      auto pick_up_pose = planTrajectory(req->pick_pose, req->robot_arm_rotation);
      if (pick_up_pose.trajectory_.joint_trajectory.points.empty())
        return;

      // sendPlanToUnity(pick_up_pose.trajectory_);
      previous_ending_joint_angles = pick_up_pose.trajectory_.joint_trajectory.points.back().positions;

      auto place_pose = planTrajectory(req->place_pose, req->robot_arm_rotation);
      if (place_pose.trajectory_.joint_trajectory.points.empty())
        return;

      sendPlanToUnity(place_pose.trajectory_);
      res->trajectories.push_back(pre_grasp_pose.trajectory_);
      res->trajectories.push_back(grasp_pose.trajectory_);
      res->trajectories.push_back(pick_up_pose.trajectory_);
      res->trajectories.push_back(place_pose.trajectory_);

      move_group_->clearPoseTargets();
    }
    catch (const std::exception &ex)
    {
      RCLCPP_ERROR(LOGGER, "Exception in planning: %s", ex.what());
    }
  }

  void sendPlanToUnity(const moveit_msgs::msg::RobotTrajectory &trajectory)
  {
    Xarm7ROSJointsToUnity msg;
    msg.trajectory = trajectory;
    pub_->publish(msg);
    RCLCPP_INFO(LOGGER, "Trajectory sent to Unity");
    is_trajectory_ = false;
  }

  std::shared_ptr<moveit::planning_interface::MoveGroupInterface> move_group_;

  rclcpp::Node::SharedPtr node_;
  rclcpp::Publisher<Xarm7ROSJointsToUnity>::SharedPtr pub_;
  rclcpp::Subscription<Xarm7UnityTargetPoseToROS>::SharedPtr sub_;
  rclcpp::Service<MoverService>::SharedPtr srv_;
  std::vector<std::string> joint_names_ = {"joint1", "joint2", "joint3", "joint4", "joint5", "joint6", "joint7"};
  bool is_trajectory_;
  moveit::planning_interface::MoveGroupInterface::Plan motion_plan_;
};

void exit_sig_handler([[maybe_unused]] int signum)
{
  fprintf(stderr, "[xarm_planner_node] Ctrl-C caught, exit process...\n");
  exit(-1);
}

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  node_options.automatically_declare_parameters_from_overrides(true);
  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("mover_node", node_options);
  auto mover_node = std::make_shared<MoverNode>(node);
  RCLCPP_INFO(node->get_logger(), "mover_node start");
  signal(SIGINT, exit_sig_handler);

  rclcpp::spin(node);
  rclcpp::shutdown();

  RCLCPP_INFO(node->get_logger(), "mover_node over");
  return 0;
}

File: xarm_custom_nodes/xarm_custom_nodes/mover.py
import rclpy
from sensor_msgs.msg import JointState
from geometry_msgs.msg import Pose

# moveit
from moveit_msgs.action import MoveGroup
 
from moveit_msgs.msg import RobotState

# Custom interfaces
from xarm_msgs.srv import MoverService

import numpy as np
import transformations

# xarm7 joints
joint_names = ["joint1", "joint2", "joint3", "joint4", "joint5", "joint6", "joint7"]



"""
    Given the start angles of the robot, plan a trajectory that ends at the destination pose.
"""
def plan_trajectory(move_group, destination_pose, start_joint_angles, robot_arm_rotation): 
    current_joint_state = JointState()
    current_joint_state.name = joint_names
    current_joint_state.position = start_joint_angles

    moveit_robot_state = RobotState()
    moveit_robot_state.joint_state = current_joint_state
    move_group.set_start_state(moveit_robot_state)

    move_group.set_pose_reference_frame('base_link')  # Assuming 'base_link' is the robot's base frame

    # Rotate the destination pose by 180 degrees
    # rotated_destination_pose = rotate_pose_180_deg(destination_pose)
    rotated_destination_pose = rotate_pose(destination_pose, robot_arm_rotation.x, robot_arm_rotation.y, robot_arm_rotation.z)
    # move_group.set_pose_target(rotated_destination_pose)
    move_group.set_pose_target(rotated_destination_pose)
    # move_group.set_pose_target(destination_pose)
    plan = move_group.plan()

    if not plan:
        exception_str = """
            Trajectory could not be planned for a destination of {}.
            Please make sure target and destination are reachable by the robot.
        """.format(destination_pose)
        raise Exception(exception_str)

    return planCompat(plan)



def rotate_pose(pose, roll, pitch, yaw):
    # Convert these angle to radians
    roll = np.deg2rad([roll])
    pitch = np.deg2rad([pitch])
    yaw = np.deg2rad([yaw])
    # Create a rotation matrix from roll, pitch, and yaw angles
    rotation_matrix = transformations.euler_matrix(roll, pitch, yaw)

    # Extract the current pose orientation as a quaternion
    current_orientation = [pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w]

    # Create the current pose transformation matrix
    current_pose_matrix = transformations.quaternion_matrix(current_orientation)
    current_pose_matrix[:3, 3] = [pose.position.x, pose.position.y, pose.position.z]

    # Apply the rotation to the current pose
    new_pose_matrix = transformations.concatenate_matrices(rotation_matrix, current_pose_matrix)

    # Extract the new position and orientation
    new_position = new_pose_matrix[:3, 3]
    new_orientation = transformations.quaternion_from_matrix(new_pose_matrix)

    # Create a new Pose object for the transformed pose
    new_pose = Pose()
    new_pose.position.x = new_position[0]
    new_pose.position.y = new_position[1]
    new_pose.position.z = new_position[2]
    new_pose.orientation.x = new_orientation[0]
    new_pose.orientation.y = new_orientation[1]
    new_pose.orientation.z = new_orientation[2]
    new_pose.orientation.w = new_orientation[3]

    return new_pose

# def rotate_pose_180_deg(pose):
#     # Create a 180 degree rotation matrix around the Z axis
#     rotation_matrix = transformations.rotation_matrix(math.pi, (0, 0, 1))

#     # Extract the current pose orientation as a quaternion
#     current_orientation = [pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w]

#     # Create the current pose transformation matrix
#     current_pose_matrix = transformations.quaternion_matrix(current_orientation)
#     current_pose_matrix[:3, 3] = [pose.position.x, pose.position.y, pose.position.z]

#     # Apply the 180 degree rotation to the current pose
#     new_pose_matrix = transformations.concatenate_matrices(rotation_matrix, current_pose_matrix)

#     # Extract the new position and orientation
#     new_position = new_pose_matrix[:3, 3]
#     new_orientation = transformations.quaternion_from_matrix(new_pose_matrix)

#     # Create a new Pose object for the transformed pose
#     new_pose = Pose()
#     new_pose.position.x = new_position[0]
#     new_pose.position.y = new_position[1]
#     new_pose.position.z = new_position[2]
#     new_pose.orientation.x = new_orientation[0]
#     new_pose.orientation.y = new_orientation[1]
#     new_pose.orientation.z = new_orientation[2]
#     new_pose.orientation.w = new_orientation[3]

#     return new_pose

"""
    Creates a pick and place plan using the four states below.
    
    1. Pre Grasp - position gripper directly above target object
    2. Grasp - lower gripper so that fingers are on either side of object
    3. Pick Up - raise gripper back to the pre grasp position
    4. Place - move gripper to desired placement position

    Gripper behaviour is handled outside of this trajectory planning.
        - Gripper close occurs after 'grasp' position has been achieved
        - Gripper open occurs after 'place' position has been achieved

    https://github.com/ros-planning/moveit/blob/master/moveit_commander/src/moveit_commander/move_group.py
"""
def plan_pick_and_place(req):
    response = MoverService.Response()

    group_name = "arm"
    move_group = moveit
    move_group = moveit_commander.MoveGroupCommander(group_name)

    current_robot_joint_configuration = req.joints_input.joints

    # Pre grasp - position gripper directly above target object
    pre_grasp_pose = plan_trajectory(move_group, req.pick_pose, current_robot_joint_configuration, req.robotArmRotation)
    
    # If the trajectory has no points, planning has failed and we return an empty response
    if not pre_grasp_pose.joint_trajectory.points:
        return response

    previous_ending_joint_angles = pre_grasp_pose.joint_trajectory.points[-1].positions

    # Grasp - lower gripper so that fingers are on either side of object
    pick_pose = copy.deepcopy(req.pick_pose)
    pick_pose.position.z -= 0.05  # Static value coming from Unity, TODO: pass along with request
    grasp_pose = plan_trajectory(move_group, pick_pose, previous_ending_joint_angles, req.robotArmRotation)
    
    if not grasp_pose.joint_trajectory.points:
        return response

    previous_ending_joint_angles = grasp_pose.joint_trajectory.points[-1].positions

    # Pick Up - raise gripper back to the pre grasp position
    pick_up_pose = plan_trajectory(move_group, req.pick_pose, previous_ending_joint_angles, req.robotArmRotation)
    
    if not pick_up_pose.joint_trajectory.points:
        return response

    previous_ending_joint_angles = pick_up_pose.joint_trajectory.points[-1].positions

    # Place - move gripper to desired placement position
    place_pose = plan_trajectory(move_group, req.place_pose, previous_ending_joint_angles, req.robotArmRotation)

    if not place_pose.joint_trajectory.points:
        return response

    # If trajectory planning worked for all pick and place stages, add plan to response
    response.trajectories.append(pre_grasp_pose)
    response.trajectories.append(grasp_pose)
    response.trajectories.append(pick_up_pose)
    response.trajectories.append(place_pose)

    move_group.clear_pose_targets()

    return response


def main(args=None):
    rclpy.init(args=args)

    mover_node = rclpy.Node("mover_node")

    rclpy.spin(mover_node)
    mover_node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
